src/DotNet/Resources/ResourceReader.cs

```diff
@@ -219,52 +220,51 @@ sealed class ResourceInfo {
 			case "System.TimeSpan": return resourceDataFactory.Create(new TimeSpan(reader.ReadInt64()));
 			case "System.Decimal":  return resourceDataFactory.Create(reader.ReadDecimal());
 			default:
-				return ReadSerializedObject(endPos, type);
+				return ReadSerializedObject(endPos, readerType, type);
 			}
 		}
 
-		IResourceData ReadSerializedObject(uint endPos, UserResourceType type) {
-			var serializedData = reader.ReadBytes((int)(endPos - reader.Position));
-			var res = createResourceDataDelegate?.Invoke(resourceDataFactory, type, serializedData);
-			return res ?? resourceDataFactory.CreateSerialized(serializedData, type);
-		}
-
-		static int ReadInt32(ref DataReader reader) {
-			try {
-				return reader.Read7BitEncodedInt32();
+		IResourceData ReadSerializedObject(uint endPos, ResourceReaderType readerType, UserResourceType type) {
+			byte[] serializedData;
+			IResourceData res;
+			switch (readerType) {
+			case ResourceReaderType.ResourceReader:
+				serializedData = reader.ReadBytes((int)(endPos - reader.Position));
+				res = createResourceDataDelegate?.Invoke(resourceDataFactory, type, serializedData, SerializationFormat.BinaryFormatter);
+				return res ?? resourceDataFactory.CreateSerialized(serializedData, SerializationFormat.BinaryFormatter, type);
+			case ResourceReaderType.DeserializingResourceReader: {
+				var format = (SerializationFormat)reader.Read7BitEncodedInt32();
```

Comment:
This enum value is read from user data so we should verify that it's a valid value before passing it to the following functions.